include/fbgemm/OutputProcessing-inl.h

```diff
@@ -44,11 +44,270 @@ inline int DoSpmdmOnInpBuffer<outT, inT, nextOPType>::f(
+  assert(
+      (B_SYMMETRIC == (Bq_zero_point_ == 0 || q_row_offsets_ == nullptr)) &&
+      "B_SYMMETRIC == true if and only if Bq_zero_point == 0 "
+      "or q_row_offsets_ == nullptr");
+  assert(
+      (HAS_BIAS == (bias_ != nullptr)) &&
+      "HAS_BIAS == true if and only if bias != nullptr");
+
+  __m256i A_zero_point_v = _mm256_set1_epi32(Aq_zero_point_);
+  __m256i C_zero_point_epi16_v = _mm256_set1_epi16(C_zero_point_);
+  __m256i C_zero_point_epi8_v = _mm256_set1_epi8(C_zero_point_);
+
+  __m256i permute_mask_v =
+      _mm256_set_epi32(0x07, 0x03, 0x06, 0x02, 0x05, 0x01, 0x04, 0x00);
+
+  constexpr int VLEN = 8;
+  for (int i = block.row_start; i < block.row_start + block.row_size; ++i) {
+    std::int32_t row_offset = (q_row_offsets_ && !B_SYMMETRIC)
+        ? q_row_offsets_[i - block.row_start] * Bq_zero_point_
+        : 0;
+    __m256i row_offset_v = _mm256_set1_epi32(row_offset);
+    int j = block.col_start;
+    for (; j < block.col_start + (block.col_size / (VLEN * 4) * (VLEN * 4));
+         j += (VLEN * 4)) {
+      __m256i x_v = _mm256_loadu_si256(reinterpret_cast<const __m256i*>(
+          inp + (i - block.row_start) * ld_in + (j - block.col_start)));
+      __m256i y_v = _mm256_loadu_si256(reinterpret_cast<const __m256i*>(
+          inp + (i - block.row_start) * ld_in + (j - block.col_start) +
+          1 * VLEN));
+      __m256i z_v = _mm256_loadu_si256(reinterpret_cast<const __m256i*>(
+          inp + (i - block.row_start) * ld_in + (j - block.col_start) +
+          2 * VLEN));
+      __m256i w_v = _mm256_loadu_si256(reinterpret_cast<const __m256i*>(
+          inp + (i - block.row_start) * ld_in + (j - block.col_start) +
+          3 * VLEN));
+
+      if (!A_SYMMETRIC) {
+        __m256i col_off_v = _mm256_mullo_epi32(
+            A_zero_point_v,
+            _mm256_loadu_si256(
+                reinterpret_cast<const __m256i*>(q_col_offsets_ + j)));
+        x_v = _mm256_sub_epi32(x_v, col_off_v);
+        col_off_v = _mm256_mullo_epi32(
+            A_zero_point_v,
+            _mm256_loadu_si256(
+                reinterpret_cast<const __m256i*>(q_col_offsets_ + j + VLEN)));
+        y_v = _mm256_sub_epi32(y_v, col_off_v);
+        col_off_v = _mm256_mullo_epi32(
+            A_zero_point_v,
+            _mm256_loadu_si256(reinterpret_cast<const __m256i*>(
+                q_col_offsets_ + j + 2 * VLEN)));
+        z_v = _mm256_sub_epi32(z_v, col_off_v);
+        col_off_v = _mm256_mullo_epi32(
+            A_zero_point_v,
+            _mm256_loadu_si256(reinterpret_cast<const __m256i*>(
+                q_col_offsets_ + j + 3 * VLEN)));
+        w_v = _mm256_sub_epi32(w_v, col_off_v);
+      }
+
+      if (!B_SYMMETRIC) {
+        x_v = _mm256_sub_epi32(x_v, row_offset_v);
+        y_v = _mm256_sub_epi32(y_v, row_offset_v);
+        z_v = _mm256_sub_epi32(z_v, row_offset_v);
+        w_v = _mm256_sub_epi32(w_v, row_offset_v);
+      }
+      if (bias_) {
```

Comment:
this should be `if (HAS_BIAS)`, as `bias_ != nullptr` is a dynamic condition, while `HAS_BIAS` is static for a given template instantiation